python/social_bot/envs/grocery_ground.py

```diff
@@ -0,0 +1,232 @@
+    it will get reward -1.
+    """
+
+    def __init__(self,
+                 max_steps=500,
+                 goal_name="bookshelf",
+                 success_distance_thresh=0.5,
+                 fail_distance_thresh=3.0,
+                 goal_random_range=10.0):
+        """
+        Args:
+            max_steps (int): episode will end if not reaching gaol in so many steps
+            goal_name (string): name of the goal in the world
+            success_distance_thresh (float): the goal is reached if it's within this distance to the agent
+            fail_distance_thresh (float): if the agent moves away from the goal more than this distance,
+                it's considered a failure and is givne reward -1
+            goal_random_range (float): the goal's random position range
+            
+        """
+        self._goal_random_range = goal_random_range
+        super(GroceryGroundGoalTask, self).__init__(
+            max_steps=max_steps,
+            goal_name=goal_name,
+            success_distance_thresh=success_distance_thresh,
+            fail_distance_thresh=fail_distance_thresh)
+
+    def _move_goal(self, goal, agent_loc):
+        while True:
+            loc = (random.random() * self._goal_random_range -
+                   self._goal_random_range / 2,
+                   random.random() * self._goal_random_range -
+                   self._goal_random_range / 2, 0)
+            self._initial_dist = np.linalg.norm(loc - agent_loc)
+            if self._initial_dist > 0.5:
+                break
+        goal.set_pose((loc, (0, 0, 0)))
+
+
+def rgb2gray(rgb):
+    return np.dot(rgb[..., :3], [0.299, 0.587, 0.114])
+
+
+class GroceryGround(gym.Env):
+    """
+    The goal of this task is to train the agent to navigate to the objects given its
+    name.
+
+    Joints of caster and arms in the PR2 are controllable by force,
+    the observations are the states of the world, including every model's position
+    and rotation.
+
+    In the version:
+    * agent will receive the total contacts of the wall as the reward
+
+    """
+
+    def __init__(self, with_language=False, use_image_obs=False, port=None):
+        """
+        Args:
+            with_language (bool): the observation will be a dict with an extra sentence
+            use_image_obs (bool): use image as observation, or use pose of the objects
+            port: Gazebo port, need to specify when run multiple environment in parallel
+        """
+        if port is None:
+            port = 0
+        gazebo.initialize(port=port)
+        self._world = gazebo.new_world_from_file(
+            os.path.join(social_bot.get_world_dir(), "grocery_ground.world"))
+        # self._world.info()
+
+        self._agent = self._world.get_agent()
+        self._pr2_joints = list(
+            filter(
+                lambda s: s.find('fl_caster_r_wheel_joint') != -1 or s.find('fr_caster_l_wheel_joint') != -1,
```

Comment:
pr2 has 4 casts: fl,fr,bl,br (front left, front right...), and each has left and right wheel.. it chould have 8 wheel joints to control moving..

Reply:
Thanks for the helpful manual. 

To simplify the caster controlling logic, I have make the front rotation joint as fixed in the model file. Now it works as a differential driver like turtle-bot and pioneer. We only need to control 2 wheel_joint in the front and leave other joints free.